AST if-statement representation has changed in recent version of Clang (9.0)

The `transpileIfStmt` logic of `c2go` relies on the assumption that if-statements always have 4 (or 5) children. With recent versions of Clang, this assumption no longer holds.

# Example

Contents of `b.c`:
```c
int f(int x) {
	if (x > 0) {
		return 42;
	}
	return 10;
}
```

# Panic

```
$ c2go transpile b.c
panic: Expected 4 children in IfStmt, got []ast.Node{(*ast.BinaryOperator)(0xc000244000), (*ast.CompoundStmt)(0xc00024c150)}

goroutine 1 [running]:
github.com/elliotchance/c2go/transpiler.transpileIfStmt(0xc00012aea0, 0xc000148a20, 0x8, 0x0, 0x4, 0xc000106d20, 0x0, 0x7, 0x0, 0x0, ...)
	/home/u/goget/src/github.com/elliotchance/c2go/transpiler/branch.go:56 +0xb8b
github.com/elliotchance/c2go/transpiler.transpileToStmt(0x70fd40, 0xc00012aea0, 0xc000148a20, 0x0, 0x0, 0x8dc308, 0x0, 0x0, 0x8dc308, 0x0, ...)
```

# AST output

The AST of `b.c` looks as follows:

```
$ c2go ast b.c
TranslationUnitDecl 0x5645b8ce69f8 <<invalid sloc>> <invalid sloc>
|-TypedefDecl 0x5645b8ce7290 <<invalid sloc>> <invalid sloc> implicit __int128_t '__int128'
| `-BuiltinType 0x5645b8ce6f90 '__int128'
|-TypedefDecl 0x5645b8ce7300 <<invalid sloc>> <invalid sloc> implicit __uint128_t 'unsigned __int128'
| `-BuiltinType 0x5645b8ce6fb0 'unsigned __int128'
|-TypedefDecl 0x5645b8ce75e8 <<invalid sloc>> <invalid sloc> implicit __NSConstantString 'struct __NSConstantString_tag'
| `-RecordType 0x5645b8ce73e0 'struct __NSConstantString_tag'
|   `-Record 0x5645b8ce7358 '__NSConstantString_tag'
|-TypedefDecl 0x5645b8ce7680 <<invalid sloc>> <invalid sloc> implicit __builtin_ms_va_list 'char *'
| `-PointerType 0x5645b8ce7640 'char *'
|   `-BuiltinType 0x5645b8ce6a90 'char'
|-TypedefDecl 0x5645b8ce7958 <<invalid sloc>> <invalid sloc> implicit __builtin_va_list 'struct __va_list_tag [1]'
| `-ConstantArrayType 0x5645b8ce7900 'struct __va_list_tag [1]' 1 
|   `-RecordType 0x5645b8ce7760 'struct __va_list_tag'
|     `-Record 0x5645b8ce76d8 '__va_list_tag'
`-FunctionDecl 0x5645b8d457d0 </home/u/Desktop/ccc/b/b.c:1:1, line:6:1> line:1:5 f 'int (int)'
  |-ParmVarDecl 0x5645b8d45700 <col:7, col:11> col:11 used x 'int'
  `-CompoundStmt 0x5645b8d459c8 <col:14, line:6:1>
    |-IfStmt 0x5645b8d45980 <line:2:2, line:4:2>
    | |-BinaryOperator 0x5645b8d45918 <line:2:6, col:10> 'int' '>'
    | | |-ImplicitCastExpr 0x5645b8d45900 <col:6> 'int' <LValueToRValue>
    | | | `-DeclRefExpr 0x5645b8d458c0 <col:6> 'int' lvalue ParmVar 0x5645b8d45700 'x' 'int'
    | | `-IntegerLiteral 0x5645b8d458e0 <col:10> 'int' 0
    | `-CompoundStmt 0x5645b8d45968 <col:13, line:4:2>
    |   `-ReturnStmt 0x5645b8d45958 <line:3:3, col:10>
    |     `-IntegerLiteral 0x5645b8d45938 <col:10> 'int' 42
    `-ReturnStmt 0x5645b8d459b8 <line:5:2, col:9>
      `-IntegerLiteral 0x5645b8d45998 <col:9> 'int' 10
```

Notice that the `IfStmt` node only has two children, the conditional node (`BinaryOperator`) and the if-body node (`CompoundStmt`).

--- a/ast/if_stmt.go
+++ b/ast/if_stmt.go
@@ -4,18 +4,20 @@ package ast
 type IfStmt struct {
 	Addr       Address
 	Pos        Position
+	HasElse    bool
 	ChildNodes []Node
 }
 
 func parseIfStmt(line string) *IfStmt {
 	groups := groupsFromRegex(
-		"<(?P<position>.*)>",
+		"<(?P<position>.*)>(?P<has_else> has_else)?",
 		line,
 	)
 
 	return &IfStmt{
 		Addr:       ParseAddress(groups["address"]),
 		Pos:        NewPositionFromString(groups["position"]),
+		HasElse:    len(groups["has_else"]) > 0,
 		ChildNodes: []Node{},
 	}
 }

--- a/ast/if_stmt_test.go
+++ b/ast/if_stmt_test.go
@@ -11,6 +11,12 @@ func TestIfStmt(t *testing.T) {
 			Pos:        NewPositionFromString("line:11:7, line:18:7"),
 			ChildNodes: []Node{},
 		},
+		`0x560d43f4ee98 <line:65:2, line:72:2> has_else`: &IfStmt{
+			Addr:       0x560d43f4ee98,
+			Pos:        NewPositionFromString("line:65:2, line:72:2"),
+			HasElse:    true,
+			ChildNodes: []Node{},
+		},
 	}
 
 	runNodeTests(t, nodes)

--- a/transpiler/branch.go
+++ b/transpiler/branch.go
@@ -24,7 +24,8 @@ func transpileIfStmt(n *ast.IfStmt, p *program.Program) (
 	postStmts := []goast.Stmt{}
 	children := n.Children()
 
-	// There is always 4 or 5 children in an IfStmt. For example:
+	// With older versions of Clang there is always 4 or 5 children in an IfStmt.
+	// For example:
 	//
 	//     if (i == 0) {
 	//         return 0;
@@ -40,6 +41,12 @@ func transpileIfStmt(n *ast.IfStmt, p *program.Program) (
 	//
 	// elseBody will be nil if there is no else clause.
 
+	// In Clang 9.0 (tested on Linux), an if-statement without an else has
+	// exactly 2 children.
+	//
+	// 1. conditional = BinaryOperator: i == 0
+	// 2. body = CompoundStmt: { return 0; }
+
 	// On linux I have seen only 4 children for an IfStmt with the same
 	// definitions above, but missing the first argument. Since we don't
 	// know what the first argument is for anyway we will just remove it on
@@ -51,19 +58,18 @@ func transpileIfStmt(n *ast.IfStmt, p *program.Program) (
 		children = children[1:]
 	}
 
-	// From here on there must be 4 children.
-	if len(children) != 4 {
-		panic(fmt.Sprintf("Expected 4 children in IfStmt, got %#v", children))
-	}
 
 	// Maybe we will discover what the nil value is?
-	if children[0] != nil {
+	if len(children) == 4 && children[0] != nil {
 		panic("non-nil child 0 in IfStmt")
+	}
+	if len(children) == 4 {
+		children = children[1:]
 	}
 
 	// The last parameter must be false because we are transpiling an
 	// expression - assignment operators need to be wrapped in closures.
-	conditional, conditionalType, newPre, newPost, err := transpileToExpr(children[1], p, false)
+	conditional, conditionalType, newPre, newPost, err := transpileToExpr(children[0], p, false)
 	if err != nil {
 		return nil, nil, nil, err
 	}
@@ -83,7 +89,7 @@ func transpileIfStmt(n *ast.IfStmt, p *program.Program) (
 		boolCondition = util.NewNil()
 	}
 
-	body, newPre, newPost, err := transpileToBlockStmt(children[2], p)
+	body, newPre, newPost, err := transpileToBlockStmt(children[1], p)
 	if err != nil {
 		return nil, nil, nil, err
 	}
@@ -101,8 +107,8 @@ func transpileIfStmt(n *ast.IfStmt, p *program.Program) (
 		Body: body,
 	}
 
-	if children[3] != nil {
-		elseBody, newPre, newPost, err := transpileToBlockStmt(children[3], p)
+	if len(children) > 2 && children[2] != nil {
+		elseBody, newPre, newPost, err := transpileToBlockStmt(children[2], p)
 		if err != nil {
 			return nil, nil, nil, err
 		}
@@ -111,7 +117,7 @@ func transpileIfStmt(n *ast.IfStmt, p *program.Program) (
 
 		if elseBody != nil {
 			r.Else = elseBody
-			if _, ok := children[3].(*ast.IfStmt); ok {
+			if _, ok := children[2].(*ast.IfStmt); ok {
 				if len(elseBody.List) == 1 {
 					r.Else = elseBody.List[0]
 				}

--- a/transpiler/branch.go
+++ b/transpiler/branch.go
@@ -46,6 +46,7 @@ func transpileIfStmt(n *ast.IfStmt, p *program.Program) (
 	//
 	// 1. conditional = BinaryOperator: i == 0
 	// 2. body = CompoundStmt: { return 0; }
+	// 3. (optional) else-body = CompoundStmt: { return 1; }
 
 	// On linux I have seen only 4 children for an IfStmt with the same
 	// definitions above, but missing the first argument. Since we don't

